### PR TITLE
[FN] Wrap period loop component stat in a Task, so that it can be cancelled.

### PR DESCRIPTION
**Pull request by fassadlr**
This is so that if a specific component fails to provide period loop stats, we will know which one is causing the issue.

This will help us debug issues like:

https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/4112

As we suspect there is a deadlock in one of the components.

### CODE REVIEW

**Comment by MithrilMan** [comment]
in my pending PR #3534 there is an extension that would simplify the cancellation of tasks with a timeout and let you await for it in a concise way
https://github.com/stratisproject/StratisBitcoinFullNode/blob/d0e54cb31c947efd51d608dcea7bb41e6e49c20a/src/Stratis.Bitcoin/Utilities/Extensions/TaskExtension.cs

basically you don't need the poilerplate code to use WaitAny on a delay task, just pass it to WithCancellationAsync

you can see it used here even if it not uses explicitly a task.delay but the concept is the same (wait the page to full load, it will move to the right position)
https://github.com/stratisproject/StratisBitcoinFullNode/pull/3534/files#diff-33aece62ca380e93ea949a0646bfa022L179

there is a link in the extension comment that you can follow to read the article where this use is explained

P.S.
as a plus, maybe consider using IAsyncProvider.RegisterTask to track the task statuses

**Comment by dangershony** [comment]
This is one big change to try to identify a deadlock 🤔

**Comment by fassadlr** [comment]
@MithrilMan thanks for the remarks on using `WithCancellationAsync`. It works brilliantly!

@dangershony all the changes are just so that the component can pass it's name to `NodeStats` :)

File: src/Stratis.Bitcoin/Utilities/NodeStats.cs [hunk 1]
```diff
@@ -62,13 +68,47 @@ public string GetStats()
...
+
+                foreach (StatsItem inlineStatItem in this.stats.Where(x => x.StatsType == StatsType.Inline))
+                {
+                    try
+                    {
+                        using (var cts = new CancellationTokenSource(TimeSpan.FromSeconds(componentStatsWaitSeconds)))
+                        {
+                            Task.Run(() =>
+                            {
+                                inlineStatItem.AppendStatsAction(statsBuilder);
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.Bitcoin/Utilities/NodeStats.cs]: OK I have added another catch for `Exception`

File: src/Stratis.SmartContracts.IntegrationTests/PoW/SmartContractMinerTests.cs [hunk 2]
```diff
@@ -187,13 +187,13 @@ public async Task InitializeAsync([CallerMemberName] string callingMethod = "")
...
-                IDateTimeProvider dateTimeProvider = DateTimeProvider.Default;
-                var inMemoryCoinView = new InMemoryCoinView(this.ChainIndexer.Tip.HashBlock);
-                this.cachedCoinView = new CachedCoinView(inMemoryCoinView, dateTimeProvider, new LoggerFactory(), new NodeStats(new DateTimeProvider()));
-
                 this.loggerFactory = new ExtendedLoggerFactory();
                 this.loggerFactory.AddConsoleWithFilters();
 
+                IDateTimeProvider dateTimeProvider = DateTimeProvider.Default;
+                var inMemoryCoinView = new InMemoryCoinView(this.ChainIndexer.Tip.HashBlock);
+                this.cachedCoinView = new CachedCoinView(inMemoryCoinView, dateTimeProvider, new LoggerFactory(), new NodeStats(dateTimeProvider, this.loggerFactory));
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.SmartContracts.IntegrationTests/PoW/SmartContractMinerTests.cs]: Fixed

File: src/Stratis.Bitcoin/Utilities/NodeStats.cs [hunk 3]
```diff
@@ -62,13 +68,47 @@ public string GetStats()
             lock (this.locker)
             {
                 string date = this.dateTimeProvider.GetUtcNow().ToString(CultureInfo.InvariantCulture);
-                statsBuilder.AppendLine($"======Node stats====== {date}");
 
-                foreach (StatsItem actionPriority in this.stats.Where(x => x.StatsType == StatsType.Inline))
-                    actionPriority.AppendStatsAction(statsBuilder);
+                // The amount of seconds the period loop will wait on a component to return its stats before cancelling.
+                int componentStatsWaitSeconds = 10;
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.Bitcoin/Utilities/NodeStats.cs]: Fixed

**Comment by fassadlr** [comment]
> about the component name, probably I'd have preferred to pass the component instance into RegisterStats and then just do a "component.GetType().Name" inside the RegisterStats method itself

I think for now we can just leave it like this. If the requirement arises to log more information from the component we can change this.

**Pull request merged**